Lib/test/test_ast/test_ast.py

```diff
@@ -135,6 +137,21 @@ def test_optimization_levels_const_folding(self):
                     res = to_tuple(tree.body[0])
                     self.assertEqual(res, expected)
 
+    def test_const_folding(self):
+        for input, folded, not_folded, kind in (
+            (eval_opt_tests, eval_results_folded, eval_results_not_folded, "eval"),
+            (exec_opt_tests, exec_results_folded, exec_results_not_folded, "exec"),
+            ):
+            for i, f, nf in zip(input, folded, not_folded):
+                with self.subTest(action="folding", input=i):
+                    ast_tree = compile(i, "?", kind, ast.PyCF_ONLY_AST | ast.PyCF_OPTIMIZED_AST)
```

Comment:
Calling `compile` with `ast.PyCF_ONLY_AST` is essentially the same as `ast.parse`, so if we generate the expected output with `ast.parse`, I don't think that actually tests the output.

Reply:
Actually, this case equal rather to ``ast.parse(i, "?", kind, optimize=1)``.

Reply:
My point was that I think it would be better to write the tests in a way that doesn't utilise the tested code to produce the expected output.

We can hard-code the expected output, or we can write an assertion that some condition hold for the actual output, I'm not sure which is easier in this case.

Reply:
The reason I don't like utilising the code to produce the expected output is what happened with the test_dis tests - there is so much "expected output" that changes, that people just update it mechanically without even looking at what changed.

Reply:
> My point was that I think it would be better to write the tests in a way that doesn't utilise the tested code to produce the expected output.
> 
> We can hard-code the expected output, or we can write an assertion that some condition hold for the actual output, I'm not sure which is easier in this case.

This sounds reasonable. I would prefer to write an assertions that "something" is(or isn't) in the actual output.

Reply:
> doesn't utilise the tested code to produce the expected output.

I don't think it's the case: expected output is hand-coded.